### PR TITLE
implement advanced spend

### LINKED ISSUE #1177: Advanced spend REST API endpoint
```
POST /wallets/transaction
Content-Type: application/json
```

```json
{
    "hours_selection": {
        "type" "manual"
    },
    "wallet": {
        "id": "foo.wlt",
        "addresses": [...],
        "password": "..."
    },
    "change_address": "...",
    "to": [{
        "address": "address",
        "coins": "123.456000",
        "hours": "123456"
    }, ...]
}
```

```json
{
    "hours_selection": {
        "type" "auto"
        "mode": "share",
        "share_factor": "0.5"
    },
    "wallet": {
        "id": "foo.wlt",
        "addresses": [...],
        "password": "..."
    },
    "change_address": "...",
    "to": [{
        "address": "address",
        "coins": "123.456000",
        "hours": "123456"
    }, ...]
}
```

"hours_selection" "type" can be "manual" or "auto".  
If "manual", all "hours" must be specified in the "to" objects.
If "auto", "hours" must not be specified in the "to" objects.

"addresses" is optional for "wallet".  If not provided, all addresses in the wallet are used.

"mode" is required for "auto" type, and must be equal to "share":

* "share" - The remaining coin hours after the burn are multiplied by "share_factor", rounded down then distributed proportionally amongst "to".

### PR DESCRIPTION
**Pull request by therealssj**
*Note*: usual spends done by all other functions usually have `change` as the first output, advanced spend puts it as the last output.

- [x] Implement advanced spend
- [ ] Tests

Fixes #1177 

Changes:
- Implement advanced spend for REST API

Does this change need to mentioned in CHANGELOG.md?
Yes

### CODE REVIEW

File: src/daemon/gateway.go [hunk 1]
```diff
@@ -529,6 +529,39 @@ func (gw *Gateway) Spend(wltID string, password []byte, coins uint64, dest ciphe
...
+			return
+		}
+
+		return
+		// Inject transaction
+		//err = gw.d.Visor.InjectBroadcastTransaction(*tx, gw.d.Pool)
+		//if err != nil {
+		//	logger.Errorf("Inject transaction failed: %v", err)
+		//	return
+		//}
```
Comment (therealssj) [inline src/daemon/gateway.go]: will uncomment this after tests are done.

File: src/gui/spend.go [hunk 2]
```diff
@@ -0,0 +1,256 @@
...
+		}
+
+		var advancedSpend wallet.AdvancedSpendRequest
+		err := json.NewDecoder(r.Body).Decode(&advancedSpend)
+		if err != nil {
+			logger.Errorf("Invalid advanced spend request: %v", err)
+			wh.Error400(w, fmt.Sprintf("Bad Request: %v", err))
+			return
+		}
+		defer func() {
```
Comment (gz-c) [inline src/gui/spend.go]: you don't need to close the request body

File: src/gui/spend.go [hunk 3]
```diff
@@ -0,0 +1,256 @@
...
+		if advancedSpend.HoursSelection.Type == "auto" {
+			switch advancedSpend.HoursSelection.Mode {
+			case "split_even":
+				switch advancedSpend.HoursSelection.ShareFactor {
+				case "":
+					logger.Error("missing hours selection share factor when mode is `split_even`")
+					wh.Error400(w, "missing hours selection share factor when mode is split_even")
+					return
+				default:
+					shareFactor, err := strconv.ParseFloat(advancedSpend.HoursSelection.ShareFactor, 64)
```
Comment (gz-c) [inline src/gui/spend.go]: use decimal.Decimal

File: src/gui/spend.go [hunk 4]
```diff
@@ -0,0 +1,256 @@
...
+
+		// create a wltmap
+		wltMap := make(map[string]struct{})
+		for _, wlt := range advancedSpend.Wallets {
+			wltMap[wlt] = struct{}{}
+		}
+
+		// fetch all wallets on the system
+		wallets, err := gateway.GetWallets()
+		if err != nil {
```
Comment (gz-c) [inline src/gui/spend.go]: there are non-500 errors like ErrWalletAPIDisabled; see the `switch err` used in other wallet handlers

File: src/gui/spend.go [hunk 5]
```diff
@@ -0,0 +1,256 @@
...
+				logger.Errorf("coinhours value missing for %v when mode is manual", to.Address)
+				wh.Error400(w, fmt.Sprintf("coinhours value missing for %v when mode is manual", to.Address))
+				return
+			}
+
+		}
+
+		// create a wltmap
+		wltMap := make(map[string]struct{})
+		for _, wlt := range advancedSpend.Wallets {
```
Comment (gz-c) [inline src/gui/spend.go]: need to check that these wallets exist

File: src/wallet/service.go [hunk 6]
```diff
@@ -390,6 +395,146 @@ func (serv *Service) CreateAndSignTransaction(wltID string, password []byte, vld
...
+
+	// calculate total coins to send
+	var totalOutCoins uint64
+	for _, to := range advancedSpend.To {
+		totalOutCoins += to.Coins
+	}
+
+	// Use the MinimizeUxOuts strategy, to use least possible uxouts
+	// this will allow more frequent spending
+	// we don't need to check whether we have sufficient balance beforehand as ChooseSpends already checks that
```
Comment (gz-c) [inline src/wallet/service.go]: this comment about ChooseSpends still valid?

File: src/wallet/service.go [hunk 7]
```diff
@@ -390,6 +395,146 @@ func (serv *Service) CreateAndSignTransaction(wltID string, password []byte, vld
...
+
+		// make sure we have enough coinhours
+		if totalAddrHours > remainingHours {
+			return nil, fee.ErrTxnInsufficientCoinHours
+		}
+
+	case "auto":
+		switch advancedSpend.HoursSelection.Mode {
+		case "split_even":
+			shareFactor, err := strconv.ParseFloat(advancedSpend.HoursSelection.ShareFactor, 64)
```
Comment (gz-c) [inline src/wallet/service.go]: decimal.Decimal

File: src/wallet/wallet.go [hunk 8]
```diff
@@ -99,6 +101,35 @@ type Options struct {
...
+type Receiver struct {
+	Address string `json:"address"`
+	Coins   string `json:"coins"`
+	Hours   string `json:"hours"`
+}
+
+type HoursSelection struct {
+	Type        string  `json:"type"`
+	Mode        string  `json:"mode"`
+	ShareFactor string `json:"share_factor"`
```
Comment (gz-c) [inline src/wallet/wallet.go]: use decimal.Decimal here so you don't have to parse it manually

File: src/wallet/wallet.go [hunk 9]
```diff
@@ -99,6 +101,35 @@ type Options struct {
 	CryptoType CryptoType // wallet encryption type, scrypt-chacha20poly1305 or sha256-xor.
 }
 
+type Receiver struct {
+	Address string `json:"address"`
+	Coins   string `json:"coins"`
+	Hours   string `json:"hours"`
```
Comment (gz-c) [inline src/wallet/wallet.go]: Implement a custom type with MarshalJSON/UnmarshalJSON for Hours and Coins, to parse string to droplets/hours respectively

File: src/wallet/wallet.go [hunk 10]
```diff
@@ -99,6 +101,35 @@ type Options struct {
...
+}
+
+type HoursSelection struct {
+	Type        string  `json:"type"`
+	Mode        string  `json:"mode"`
+	ShareFactor string `json:"share_factor"`
+}
+
+type AdvancedSpendRequest struct {
```
Comment (gz-c) [inline src/wallet/wallet.go]: these json structs are better off in the gui library, then convert to the struct used by wallet as argument

File: src/wallet/service.go (hunk 6 above)
Comment (therealssj) [reply src/wallet/service.go]: yes.
basically what I meant was that I do not need to iterate over `uxb` and calculate sum of coins in all the addresses beforehand and compare with `totalOutCoins` as `ChooseSpends` handles that

File: src/wallet/wallet.go (hunk 8 above)
Comment (gz-c) [reply src/wallet/wallet.go]: use a pointer

Comment (therealssj) [reply src/wallet/wallet.go]: oh yea

File: src/gui/spend.go (hunk 4 above)
Comment (therealssj) [reply src/gui/spend.go]: done

**Comment by gz-c** [comment]
moved to #1223

**Pull request closed without merging**